### PR TITLE
executor, planner: add support for SQL_CALC_FOUND_ROWS

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/21786

Problem Summary:

The feature `SQL_CALC_FOUND_ROWS` is deprecated in MySQL, but it is unfortunately used in a number of common apps like wordpress and drupal. This causes compatibility issues with TiDB, and can leave a bad impression because such 'simple' applications fail.

The implementation disables all push-down to the coprocessor and counts rows inside TiDB. The performance requirements of this feature are not important, since it is deprecated and only provided for compatibility purposes.

### What is changed and how it works?

What's Changed:

Support for SELECT SQL_CALC_FOUND_ROWS has been added for MySQL compatibility. However, where possible it is recommended to send 2 queries to TiDB instead.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Increased code complexity.

### Release note 

- Support for SELECT SQL_CALC_FOUND_ROWS has been added for MySQL compatibility. However, where possible it is recommended to send 2 queries to TiDB instead.

### CODE REVIEW

File: executor/executor.go [hunk 1]
```diff
@@ -1156,6 +1163,70 @@ func (e *LimitExec) adjustRequiredRows(chk *chunk.Chunk) *chunk.Chunk {
 	return chk.SetRequiredRows(mathutil.Min(limitTotal, limitRequired), e.maxChunkSize)
 }
 
+func minUint64(a, b uint64) uint64 {
```
Comment (bb7133) [inline executor/executor.go:1166]: You can use `mathutil.MinUint64` instead.

Comment (morgo) [reply executor/executor.go:1166]: Nice, thanks!

**Review by bb7133 — COMMENTED** [review]
Please consider the following case:

```
create table t2(a int, b int);
insert into t2 values (1, 1), (2, 2);
select SQL_CALC_FOUND_ROWS b,count(*) as c from t2 group by b order by c limit 1;
select found_rows();
```
The expected result is 2 but TiDB with this PR gives 1.

**Comment by morgo** [comment]
@bb7133 This issue has been addressed. PTAL again :-)

**Comment by bb7133** [comment]
Unfortunately, some new cases are found:

```
DROP TABLE t1;
#
# Bug #17833261 FOUND_ROWS(): DIFFERENT RESULTS FOR LOCAL LIMIT WITH AND
# WITHOUT OFFSET IN UNION
#
CREATE TABLE t1 (i INT);
INSERT INTO t1 VALUES (1), (2), (3), (4), (5);
(SELECT SQL_CALC_FOUND_ROWS * FROM t1 LIMIT 2 OFFSET 2)
UNION ALL
(SELECT 1 FROM t1 WHERE FALSE);
i
3
4
SELECT FOUND_ROWS();
```
`2` is expected but this PR gives `4`

**Comment by bb7133** [comment]
Another edge case is about `LIMIT 0`:

```
SELECT SQL_CALC_FOUND_ROWS 'foo' LIMIT 0;
Empty set (0.00 sec)

SELECT FOUND_ROWS();
```
`1` is expected but this PR gives `0`.

**Comment by morgo** [comment]
I have changed this so that only the simple cases are supported, with no unions, select from dual or queries without a LIMIT.

@bb7133 PTAL again

**Review by bb7133 — APPROVED** [review]

**Comment by bb7133** [comment]
LGTM, /cc @qw4990 @eurekaka @XuHuaiyu

**Comment by dveeden** [comment]
Can you link tot the docs PR?

**Comment by dveeden** [comment]
Testing this a bit confused me a great deal.

With TiDB:
```
tidb SELECT * FROM t1;
+---+
| i |
+---+
| 1 |
| 2 |
| 3 |
| 4 |
| 5 |
+---+
5 rows in set (0.0005 sec)

tidb> SELECT * FROM t1 LIMIT 2;
+---+
| i |
+---+
| 1 |
| 2 |
+---+
2 rows in set (0.0006 sec)

tidb> SELECT SQL_CALC_FOUND_ROWS * FROM t1 LIMIT 2;
+---+
| i |
+---+
| 1 |
| 2 |
+---+
2 rows in set (0.0008 sec)

tidb> SELECT FOUND_ROWS();
+--------------+
| FOUND_ROWS() |
+--------------+
|            1 |
+--------------+
1 row in set (0.0008 sec)
```

This didn't match the description of this MR or what I would expect to happen.

So testing with MySQL:
```
mysql> CREATE TABLE t1 (i INT);
Query OK, 0 rows affected (0.0720 sec)

mysql> INSERT INTO t1 VALUES (1), (2), (3), (4), (5);
Query OK, 5 rows affected (0.0221 sec)

Records: 5  Duplicates: 0  Warnings: 0

mysql> SELECT * FROM t1;
+---+
| i |
+---+
| 1 |
| 2 |
| 3 |
| 4 |
| 5 |
+---+
5 rows in set (0.0004 sec)

mysql> SELECT * FROM t1 LIMIT 2;
+---+
| i |
+---+
| 1 |
| 2 |
+---+
2 rows in set (0.0003 sec)

mysql> SELECT SQL_CALC_FOUND_ROWS * FROM t1 LIMIT 2;
+---+
| i |
+---+
| 1 |
| 2 |
+---+
2 rows in set, 1 warning (0.0004 sec)
Warning (code 1287): SQL_CALC_FOUND_ROWS is deprecated and will be removed in a future release. Consider using two separate queries instead.

mysql> SELECT FOUND_ROWS();
+--------------+
| FOUND_ROWS() |
+--------------+
|            1 |
+--------------+
1 row in set, 1 warning (0.0002 sec)
Warning (code 1287): FOUND_ROWS() is deprecated and will be removed in a future release. Consider using COUNT(*) instead.
```

This also didn't match the behavior I would expect.

Also note the warnings that MySQL emits. @morgo could we add warnings like that to TiDB?

What turned out to be the issue is that the behavior with MySQL Shell (at least with a classic session) is different than what happens with MySQL Client for some reason.

**Comment by dveeden** [comment]
Another issue I see is this:
```
mysql> SELECT SQL_CALC_FOUND_ROWS * FROM (SELECT 1) a LIMIT 2;
+---+
| 1 |
+---+
| 1 |
+---+
1 row in set (0.00 sec)

mysql> SELECT SQL_CALC_FOUND_ROWS 1 LIMIT 2;
ERROR 1235 (42000): function SQL_CALC_FOUND_ROWS has only noop implementation in tidb now, use tidb_enable_noop_functions to enable these functions
```

Only allowing `SQL_CALC_FOUND_ROWS` in a certain context is fine, but the error doesn't indicate this. I assume this is because this is a generic error message? Could we make the error message more specific and/or make sure this is clearly stated in the docs? Could/should we emit a warning if the noop implementation of `SQL_CALC_FOUND_ROWS` was used?

**Comment by dveeden** [comment]
I filed https://bugs.mysql.com/bug.php?id=104502 for this.

**Comment by heipark** [comment]
@morgo @bb7133 
How do you think @dveeden  said above ?

**Comment by morgo** [comment]
> @morgo @bb7133
> How do you think @dveeden said above ?

Feedback LGTM. I am working on this when I have spare time, so it will take me a couple of days to get back to it.

**Comment by morgo** [comment]
> mysql> SELECT SQL_CALC_FOUND_ROWS * FROM t1 LIMIT 2;
> +---+
> | i |
> +---+
> | 1 |
> | 2 |
> +---+
> 2 rows in set, 1 warning (0.0004 sec)
> Warning (code 1287): SQL_CALC_FOUND_ROWS is deprecated and will be removed in a future release. Consider using two separate queries instead.
> 
> mysql> SELECT FOUND_ROWS();
> +--------------+
> | FOUND_ROWS() |
> +--------------+
> |            1 |
> +--------------+
> 1 row in set, 1 warning (0.0002 sec)
> Warning (code 1287): FOUND_ROWS() is deprecated and will be removed in a future release. Consider using COUNT(*) instead.
> ```
> 
> This also didn't match the behavior I would expect.
> 
> Also note the warnings that MySQL emits. @morgo could we add warnings like that to TiDB?

What do you think about adding the warning for `SQL_CALC_FOUND_ROWS`, but not the warning for `FOUND_ROWS()`? My rationale is that these warnings are not present in MySQL 5.7, which is the current compatibility target.

My concern is that some clients use `FOUND_ROWS()` in a benign way unassociated with `SQL_CALC_FOUND_ROWS`, and we will be breaking 5.7 compatibility for them.

> What turned out to be the issue is that the behavior with MySQL Shell (with both a classic session and an X Protocol session) is different than what happens with MySQL Client for some reason.

Could this be related to the `CLIENT_FOUND_ROWS` [protocol capability](https://dev.mysql.com/doc/internals/en/capability-flags.html)? The test case you provided is working for me. If you can produce a different result please let me know the client and if you are using tikv or not.

**Review by dveeden — APPROVED** [review]

**Comment by dveeden** [comment]
To clarify things: With MySQL Shell 8.0.x and either MySQL 8.0 or TiDB with this PR both give results that are wrong. I filed a bug for this with Oracle and it was verified. With MySQL Client 8.0.x things work ok with both MySQL and TIDB.

And with the updates to warnings this LGTM

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -1833,8 +1833,11 @@ func extractLimitCountOffset(ctx sessionctx.Context, limit *ast.Limit) (count ui
 	return count, offset, nil
 }
 
-func (b *PlanBuilder) buildLimit(src LogicalPlan, limit *ast.Limit) (LogicalPlan, error) {
-	b.optFlag = b.optFlag | flagPushDownTopN
+func (b *PlanBuilder) buildLimit(src LogicalPlan, limit *ast.Limit, calcFoundRows bool) (LogicalPlan, error) {
+	if !calcFoundRows {
+		// Only pushdown limit when calcFoundRows is not set.
+		b.optFlag = b.optFlag | flagPushDownTopN
+	}
```
Comment (eurekaka) [inline planner/core/logical_plan_builder.go:1840]: We should prevent the limit from being pushed down to the coprocessor as well? and we should not embed the limit into `IndexLookUp` either, please see `sinkIntoIndexLookUp()` for details.

**Review by eurekaka — COMMENTED** [review]
I think this feature is pretty flaky, and is inclined to produce unexpected query results. We need to be extremely careful, especially in the executor.

**Pull request closed without merging**